include/nccl_ofi_sendrecv.h

```diff
@@ -50,36 +50,70 @@ class nccl_net_ofi_sendrecv_mr_handle_t : public nccl_net_ofi_mr_handle_t {
 	ofi_mr_ptr mr;
 };
 
-typedef struct nccl_net_ofi_sendrecv_listen_comm {
-	/* This base listen communicator must be the first member of
-	 * this struct. This allows casting between pointers of this
-	 * struct and its base struct. */
-	nccl_net_ofi_listen_comm_t base;
+/* Forward declarations needed for sendrecv transport endpoint and comm types */
+class nccl_net_ofi_sendrecv_device_t;
+class nccl_net_ofi_sendrecv_ep_t;
+
+class nccl_net_ofi_sendrecv_listen_comm_t : public nccl_net_ofi_listen_comm_t {
+public:
+	/**
+	 * @brief Return SENDRECV listen communicator endpoint
+	 */
+	inline nccl_net_ofi_sendrecv_ep_t *sendrecv_listen_comm_get_ep()
+	{
+		return reinterpret_cast<nccl_net_ofi_sendrecv_ep_t *>(ep);
+	}
 
-	struct fid_ep *local_ep;
+	int accept(nccl_net_ofi_xfer_comm_t **recv_comm) override;
+
+	int close() override;
+
+	struct fid_ep *local_ep = nullptr;
 	fi_addr_t local_ep_addr;
 	/* Saves temporary state when creating receive communicator object */
 	save_comm_state_t state;
 
-	nccl_ofi_cm_listener *listener;
-} nccl_net_ofi_sendrecv_listen_comm_t;
+	nccl_ofi_cm_listener *listener = nullptr;
+};
 
-typedef struct nccl_net_ofi_sendrecv_send_comm {
-	/* This base send communicator must be the first member of this
-	 * struct. This allows casting between pointers of this struct
-	 * and its base struct. */
-	nccl_net_ofi_send_comm_t base;
+class nccl_net_ofi_sendrecv_send_comm_t : public nccl_net_ofi_xfer_comm_t {
+public:
+	/**
+	 * @brief Return SENDRECV send communicator endpoint
+	 */
+	inline nccl_net_ofi_sendrecv_ep_t *sendrecv_send_comm_get_ep()
+	{
+		return reinterpret_cast<nccl_net_ofi_sendrecv_ep_t *>(ep);
+	}
+
+	int send(void *data, size_t size, int tag,
+		 nccl_net_ofi_mr_handle_t *mhandle, nccl_net_ofi_req_t **req) override;
+
+	int close() override;
+
+	/* write and write_inline operation not support for SENDRECV protocol */
+	int write(void *src, size_t size, void *src_mhandle, uint64_t dest,
+		  uint64_t mr_key, nccl_net_ofi_req_t **req) override
+	{
+		return log_unsupported_operation("protocol", "iwrite", -ENOTSUP);
+	}
+
+	int write_inline(void *src, size_t size, uint64_t dest, uint64_t mr_key,
+			 nccl_net_ofi_req_t **request) override
+	{
+		return log_unsupported_operation("protocol", "iwrite_inline", -ENOTSUP);
+	}
 
 	uint64_t num_inflight_reqs;
-	nccl_ofi_freelist_t *nccl_ofi_reqs_fl;
+	nccl_ofi_freelist_t *nccl_ofi_reqs_fl = nullptr;
 
 	uint64_t tag;
 	fi_addr_t remote_ep;
 	fi_addr_t local_ep_addr;
-	struct fid_ep *local_ep;
+	struct fid_ep *local_ep = nullptr;
```

Comment:
Maybe have declare the `struct fid_ep` type using `using` keyword elsewhere and use the type here instead of explicitly using `struct` along with it.